### PR TITLE
[flink] support one or multiple databases in a dedicated compaction job with one combined source and sink

### PR DESCRIPTION
**Pull request by LsomeYeah**
### Purpose

Linked issue: close #xxx

It would be beneficial for Paimon to support compaction across multiple tables in a single job. Specially, it should have the ability to detect newly created tables after the running compaction job.

This pr support：

- support one or multiple databases compaction in a dedicated job with one combined source and combined sink.
- support detecting newly created tables after the running compaction job
- support including-tables.
- support excluding-tables.

### Tests

`org.apache.paimon.flink.source.MultiTablesCompactorSourceBuilderITCase`
This is for testing if the combined source can work normally.

`org.apache.paimon.flink.action.MultiTablesCompactActionITCase`
This is for testing if the compaction job for multi-tables  can work normally.

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java [hunk 1]
```diff
@@ -54,13 +56,29 @@ public class StoreMultiCommitter
...
+    public StoreMultiCommitter(
+            Catalog.Loader catalogLoader,
+            String commitUser,
+            @Nullable CommitterMetrics metrics,
+            boolean isCompactJob) {
+        this.catalog = catalogLoader.load();
+        this.commitUser = commitUser;
+        this.metrics = metrics;
+        this.tableCommitters = new HashMap<>();
+        this.isCompactJob = isCompactJob;
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java:81]: Use `this(...)`

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java [hunk 2]
```diff
@@ -148,15 +166,27 @@ private StoreCommitter getStoreCommitter(Identifier tableId) {
...
-                            table.newCommit(commitUser).ignoreEmptyCommit(false), metrics);
+            if (isCompactJob) {
+                committer =
+                        new StoreCommitter(
+                                table.newCommit(commitUser).ignoreEmptyCommit(true), metrics);
+            } else {
+                committer =
+                        new StoreCommitter(
+                                table.newCommit(commitUser).ignoreEmptyCommit(false), metrics);
+            }
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java:189]: `table.newCommit().ignoreEmptyCommit(isCompactJob)`

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/MultiTablesCompactorSourceBuilder.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public MultiTablesCompactorSourceBuilder withEnv(StreamExecutionEnvironment env) {
+        this.env = env;
+        return this;
+    }
+
+    public DataStream<RowData> build() {
+        if (env == null) {
+            throw new IllegalArgumentException("StreamExecutionEnvironment should not be null.");
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/MultiTablesCompactorSourceBuilder.java:74]: checkArgument

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MultiTablesReadOperator.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+/** this is a doc. */
+public class MultiTablesReadOperator extends AbstractStreamOperator<RowData>
+        implements OneInputStreamOperator<Tuple2<Split, String>, RowData> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final Catalog.Loader catalogLoader;
+    private final Pattern includingPattern;
+    private final Pattern excludingPattern;
+    private final Pattern databasePattern;
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MultiTablesReadOperator.java:61]: The upstream should make sure the data is valid.

File: paimon-core/src/main/java/org/apache/paimon/table/system/BucketsMultiTable.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.paimon.utils.SerializationUtils.newBytesType;
+import static org.apache.paimon.utils.SerializationUtils.serializeBinaryRow;
+
+/** this is a doc. */
+public class BucketsMultiTable implements DataTable, ReadonlyTable {
```
[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-core/src/main/java/org/apache/paimon/table/system/BucketsMultiTable.java]: done

[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-core/src/main/java/org/apache/paimon/table/system/BucketsMultiTable.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/DatabaseCompactMode.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.paimon.flink.action;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+
+/** this is a doc. */
+public enum DatabaseCompactMode implements Serializable {
```
[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/DatabaseCompactMode.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/CompactDatabaseAction.java [hunk 7]
```diff
@@ -53,21 +58,37 @@ public class CompactDatabaseAction extends ActionBase {
 
     private final Pattern includingPattern;
     private final Pattern excludingPattern;
```
[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/CompactDatabaseAction.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/MultiTablesCompactorSinkBuilder.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.paimon.options.Options;
+
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.table.data.RowData;
+
+import static org.apache.paimon.flink.sink.FlinkStreamPartitioner.partition;
+
+/** this is a doc. */
+public class MultiTablesCompactorSinkBuilder {
```
[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/MultiTablesCompactorSinkBuilder.java]: `MultiTablesCompactorSinkBuilder` has been removed and  `MultiTablesCompactorSink` is built in `CompactDatabaseAction` now.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java (hunk 1 above)
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java (hunk 2 above)
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreMultiCommitter.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/MultiTablesCompactorSourceBuilder.java (hunk 3 above)
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/MultiTablesCompactorSourceBuilder.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MultiTablesReadOperator.java (hunk 4 above)
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MultiTablesReadOperator.java]: The step of initializing table map in `MultiTablesReadOperator` has been removed.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/MultiTablesCompactorSink.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+import java.io.Serializable;
+import java.util.UUID;
+
+import static org.apache.paimon.flink.FlinkConnectorOptions.SINK_MANAGED_WRITER_BUFFER_MEMORY;
+import static org.apache.paimon.flink.FlinkConnectorOptions.SINK_USE_MANAGED_MEMORY;
+import static org.apache.paimon.utils.Preconditions.checkArgument;
+
+/** this is a doc. */
+public class MultiTablesCompactorSink implements Serializable {
```
[the remark this answers is not in the document]
Comment (LsomeYeah) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/MultiTablesCompactorSink.java:52]: Waiting for refactoring `FlinkSink`.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**